[kbn-mcp-dev-server] Refactor MCP Dev Server

simianhacker opened the pull request:
## 🍒 Summary

This commit refactors the way tools are loaded and registered in the `@kbn/mcp-dev-server` package.

The previous implementation required exporting a tool's input schema, type, and logic separately. The new approach encapsulates these into a single `ToolDefinition` object, which simplifies the process of adding and managing tools.

### 🔑 Key changes  

- Introduces a `ToolDefinition` type and an `addTool` helper function to standardize tool registration.
- Refactors the existing tools (`list_packages`, `generate_package`, `list_teams`, `run_unit_tests`) to use the new `ToolDefinition` format.
- Updates the server to use the `addTool` helper, cleaning up the registration logic.
- Restores and updates the Jest tests for all refactored tools to ensure correctness.
- Adds a new test for the `addTool` helper itself.
- Updates the `README.md` to reflect the new, simpler process for
  adding tools to the server.

## 🚀 Adding a New Tool to the MCP Dev Server

This document outlines the process of adding a new tool to the MCP dev server.

### 1. Create a Tool File

Create a new file for your tool inside `src/platform/packages/shared/kbn-mcp-dev-server/src/tools`. The file name should be descriptive of the tool's functionality (e.g., `my_new_tool.ts`).

### 2. Implement the Tool Definition

Inside the new file, create a `ToolDefinition` object. This object will contain the tool's name, description, input schema, and handler function.

**Example:** `src/platform/packages/shared/kbn-mcp-dev-server/src/tools/my_new_tool.ts`

```typescript
import { z } from '@kbn/zod';
import { ToolDefinition } from '@kbn/mcp-server-common';

// Define the input schema for your tool using Zod
const myNewToolInputSchema = z.object({
  myArgument: z.string().describe("A description for this argument"),
});

// Implement the tool's logic in a function
async function myNewTool(input: z.infer<typeof myNewToolInputSchema>): Promise<string> {
  // Your tool's logic goes here
  return `You passed the argument: ${input.myArgument}`;
}

// Create the tool definition
export const myNewTool: ToolDefinition<typeof myNewToolInputSchema> = {
  name: 'my_new_tool',
  description: 'A brief description of what my new tool does.',
  inputSchema: myNewToolInputSchema,
  handler: async (input) => {
    const result = await myNewTool(input);
    return {
      content: [
        {
          type: 'text',
          text: result,
        },
      ],
    };
  },
};
```

### 3. Register the Tool

Open `src/platform/packages/shared/kbn-mcp-dev-server/src/server/index.ts` and follow these steps:

#### a. Import the Tool

Add an import statement at the top of the file to import your new tool's definition.

```typescript
import { myNewTool } from '../tools/my_new_tool';
```

#### b. Register the Tool with the Server

Use the `addTool` helper function to register your new tool.

```typescript
addTool(server, myNewTool);
```

🤖 This PR was assisted by Gemini CLI

simianhacker committed:
Refactor MCP dev server tool loading

- Refactored the tool loading mechanism in the `@kbn/mcp-dev-server` package to use a unified `ToolDefinition` object.
- Created a new `addTool` helper function to simplify and standardize tool registration.
- Updated all existing tools to the new `ToolDefinition` format.
- Restored and updated Jest tests for all refactored tools and added a new test for the `addTool` helper.
- Corrected all file path and type errors that arose during the refactoring process.
- Updated the `README.md` to reflect the new process for adding tools.

Prompts:

- "I would like to refactor how the tools are loaded into the kbn-mcp-dev-server package. Instead of exporting the input schema, input type, and tool's logic. I would like the tool to export an objct like:
```
export const myNewToolDef = {
  name: 'my_new_tool',
  description: 'A brief description of waht my new tool does',
  inputSchema: z.object({
    query: z.string().describe('A description for this argument'),
  }),
  handler: async (input) => {
    const result = await myNewTool(input);
    return {
      content: [
        {
          type: 'text',
          text: result,
        },
      ],
    };
  }
}
```

then in the server it's registered

```
import { myNewToolDef } from './tools/my_new_tool';

addTool(server, myNewToolDef);

```

Impliment the `addTool` helper function and refactor the existing tools. Also create a type definition for the addTool function and the tool definition in `src/types`. Create a plan for this and await my approval before executing."
- "Add a `src/utils` directory to the project and put `add_tool.ts` with a barrel export."
- "One more thing, update the README for adding a new tool to reflect this new approach."
- "for the zod package use `@kbn/zod`"
- "It looks like you messed the paths up for where you put the `utils` and `types`. You put them under `packages/kbn-mcp-dev-server` and `packages/kbn-mcp-server-common`. Those should have gone under `src/platform/packages/shared/kbn-mcp-dev-server` with the rest of the code."
- "I'm getting a type error for line 12 in the tool_definition.ts type. "Unable to resolve import""
- "It doesn't appear that any of the types are include in the modelcontextprotocol sdk."
- "you're implimentation of utils/add_tool.ts is wrong..."
- "I'm getting this type error in add_tool.ts:
Type 'T' is not assignable to type 'ZodRawShape | undefined'.
  Type 'ZodObject<any, any, any, any, any>' is not assignable to type 'ZodRawShape'.
    Type 'T' is not assignable to type 'ZodRawShape'.
      Type 'ZodObject<any, any, any, any, any>' is not assignable to type 'ZodRawShape'.
        Index signature for type 'string' is missing in type 'ZodObject<any, any, any, any, any>'.

Related information:

  * mcp.d.ts#106,9: The expected type comes from property 'inputSchema' which is declared here on type '{ title?: string | undefined; description?: string | undefined; inputSchema?: ZodRawShape | undefined; outputSchema?: ZodRawShape | undefined; annotations?: { ...; } | undefined; }'

 (tsserver 2322)"
- "I think the correct response type (instead of our ToolResponse) is `CallToolResult` from the modelcontextprotocol package."
- "It looks like you deleted the tests for each of the tools."
- "Can you add Jest test for add_tool.ts?"
- "Can you typecheck again"

🤖 This commit was assisted by Gemini CLI

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/refactor-mcp-dev-server

kibanamachine committed:
[CI] Auto-commit changed files from 'ts-node .buildkite/pipeline-resource-definitions/scripts/fix-location-collection.ts'

qn895 commented on src/platform/packages/shared/kbn-mcp-dev-server/src/server/index.ts:
```diff
@@ -10,111 +10,20 @@
-      };
-    }
-  );
+  addTool(server, listKibanaPackagesTool);
```
Nice!

qn895 commented on src/platform/packages/shared/kbn-mcp-dev-server/README.md:
```diff
@@ -14,9 +14,9 @@ node scripts/mcp_dev
+## MCP configuration example
 
-Add the following snippet to your `~/.codeium/windsurf/mcp_config.json` to let Windsurf automatically start/attach to the dev server (it assumes you use `nvm`):
+Add the following snippet to your settings for your AI Coding ide so it will automatically start/attach to the dev server (it assumes you use `nvm`):
```
```suggestion
Add the following snippet to your settings for your AI Coding IDE so it will automatically start/attach to the dev server (it assumes you use `nvm`):
```

qn895 reviewed (approved):
LGTM

simianhacker committed:
Fix failing test in add_tool.test.ts

- Updated the jest.mock path in `add_tool.test.ts` to correctly match the import path for `@modelcontextprotocol/sdk/server/mcp.js`.

Prompts:

- "I need to debug a test `yarn test:jest src/platform/packages/shared/kbn-mcp-dev-server/src/utils/add_tool.test.ts`"
- "commit changes"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Update src/platform/packages/shared/kbn-mcp-dev-server/README.md

Co-authored-by: Quynh Nguyen (Quinn) <43350163+qn895@users.noreply.github.com>

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/refactor-mcp-dev-server
